### PR TITLE
Add custom assertions fields to app models

### PR DESCRIPTION
**Pull request by esoergel**
## Technical Summary
https://docs.google.com/document/d/112St-i7h7jZHpqxcAU-d6VASftXKLfiFnqkS3ovsEVU/edit
https://dimagi-dev.atlassian.net/browse/USH-2905
This PR simply adds the model changes necessary for the custom assertions work.  I wanted to try splitting this out to head off any potential issues of deploying things on staging, possibly reverting and causing `wrap` errors, and to enable multiple people to work on it if we go that way.

### Safety story
Adds new fields to the model, no other changes.

### QA Plan

None yet

### Migrations

There isn't a migration, but there is a new field added to a couch model.  It can be rolled out whenever and does not introduce any backwards incompatabilities.

### Rollback instructions

I believe this can be rolled back without issue, but it's an isolated set of changes specifically so a rollback on the other work need not consider the couch schema changing.

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/app_manager/models.py [hunk 1]
```diff
@@ -2226,6 +2226,7 @@ class ModuleBase(IndexedSchema, ModuleMediaMixin, NavMenuItemMediaMixin, Comment
     auto_select_case = BooleanProperty(default=False)
     is_training_module = BooleanProperty(default=False)
     session_endpoint_id = StringProperty(exclude_if_none=True)  # See toggles.SESSION_ENDPOINTS
+    custom_assertions = SchemaListProperty(CustomAssertion)
```
Comment (esoergel) [inline corehq/apps/app_manager/models.py:2229]: Reviewers: I'd love your take on whether to add this to `ModuleBase`, as I've done here, versus `Module`.  This seemed to make the most sense, as conceptually, assertions could logically apply to `ReportModule`s and `ShadowModule`s too, whether or not we make that possible in the UI

Comment (orangejenny) [reply corehq/apps/app_manager/models.py:2229]: What do you think are the odds a shadow module would want its own assertion, versus inheriting from its source module?

Comment (snopoke) [reply corehq/apps/app_manager/models.py:2229]: Conceptually this seems like the right place to me though if we don't implement it for other module types it's a bit weird to have it in the model. If it's not hard to implement for all module types I would just do that (perhaps with the exception of ShadowModules - at least for now)

**Review by snopoke — COMMENTED** [review]
Are you also going to add it at the App level?

**Comment by esoergel** [comment]
> Are you also going to add it at the App level?

@snopoke do you mean in this PR?  That's already in there - this adds it to `ModuleBase` and `Application`.  Or if you mean in the UI, then yes I intend to add it in the UI for modules and apps, in addition to forms, where it's already present

File: corehq/apps/app_manager/models.py (hunk 1 above)
Comment (esoergel) [reply corehq/apps/app_manager/models.py:2229]: I think it should be possible to do it for all module types - it looks like they use a shared template for the settings page (except for I think report modules).

> What do you think are the odds a shadow module would want its own assertion, versus inheriting from its source module?

I was initially thinking shadow module assertions would be added on top of any assertions in its source module, but now I'm not so sure.  If the assertions are present to encode assumptions used in the module's forms, then yes inheriting from the source module makes sense.  However if assertions are there to assist with say case list, detail, or search, then having their own assertions makes sense.

Where I'm landing now is that shadow modules should get their own assertions, and not inherit from the source module.  Maybe it's a cop-out, but it doesn't make any assumptions about the purpose about the assertions, and allows the most flexibility.  If source module assertions _should_ apply, app builders can always copy them over (or move them to the app level, if appropriate).

**Comment by snopoke** [comment]
> > Are you also going to add it at the App level?
> 
> @snopoke do you mean in this PR? That's already in there - this adds it to `ModuleBase` and `Application`. Or if you mean in the UI, then yes I intend to add it in the UI for modules and apps, in addition to forms, where it's already present

Ah, I see - I just missed that when looking at the diff.

**Review by snopoke — APPROVED** [review]

**Comment by esoergel** [comment]
Tests didn't run on the last commit, I merged in master to nudge them

**Review by snopoke — APPROVED** [review]

**Pull request merged**